scipy/stats/tests/test_sampling.py

```diff
@@ -0,0 +1,619 @@
+from scipy.stats import UNURANError
+from scipy import stats
+from scipy.stats import chisquare, cramervonmises
+from scipy.stats._distr_params import distdiscrete
+
+
+# common test data: this data can be shared between all the tests.
+
+
+# Normal distribution shared between all the continuous methods
+class common_cont_dist:
+    # Private methods are used for performance. Public methods do a lot of
+    # validations and expensive numpy operations (like broadcasting), hence,
+    # slowing down some of the methods significantly. We can use private
+    # methods as we know that the data is guaranteed to be inside the domain
+    # and a scalar (so, no masking bad values and broadcasting required).
+    pdf = stats.norm._pdf
+    dpdf = lambda x: -x * stats.norm._pdf(x)  # noqa: E731
+    cdf = stats.norm._cdf
+
+
+# A binomial distribution to share between all the discrete methods
+class common_discr_dist:
+    params = (10, 0.2)
+    pmf = stats.binom._pmf
+    cdf = stats.binom._cdf
+
+
+all_methods = [
+    ("TransformedDensityRejection", {"dist": common_cont_dist}),
+    ("DiscreteAliasUrn", {"pv": [0.02, 0.18, 0.8]})
+]
+
+
+bad_pdfs_common = [
+    # Negative PDF
+    (lambda x: -x, UNURANError, r"50 : PDF\(x\) < 0.!"),
+    # Returning wrong type
+    (lambda x: [], TypeError, r"must be real number, not list"),
+    # Undefined name inside the function
+    (lambda x: foo, NameError, r"name 'foo' is not defined"),  # type: ignore[name-defined]  # noqa
+    # Infinite value returned => Overflow error.
+    (lambda x: np.inf, UNURANError, r"50 : PDF\(x\) overflow"),
+    # NaN value => internal error in UNU.RAN
+    # Currently, UNU.RAN just returns a "cannot create bounded hat!"
+    # error which is not very useful. So, instead of testing the error
+    # message, we just ensure an error is raised.
+    (lambda x: np.nan, UNURANError, r"..."),
+    # signature of PDF wrong
+    (lambda: 1.0, TypeError, r"takes 0 positional arguments but 1 was given")
+]
+
+bad_dpdf_common = [
+    # Infinite value returned. For dPDF, UNU.RAN complains that it cannot
+    # create a bounded hat instead of an overflow error.
+    (lambda x: np.inf, UNURANError, r"51 : cannot create bounded hat"),
+    # NaN value => internal error in UNU.RAN
+    # Currently, UNU.RAN just returns a "cannot create bounded hat!"
+    # error which is not very useful. So, instead of testing the error
+    # message, we just ensure an error is raised.
+    (lambda x: np.nan, UNURANError, r"..."),
+    # Returning wrong type
+    (lambda x: [], TypeError, r"must be real number, not list"),
+    # Undefined name inside the function
+    (lambda x: foo, NameError, r"name 'foo' is not defined"),  # type: ignore[name-defined]  # noqa
+    # signature of dPDF wrong
+    (lambda: 1.0, TypeError, r"takes 0 positional arguments but 1 was given")
+]
+
+# XXX: currently this list is not being used in any of the test. But as we
```

Comment:
> we can discuss it later in our call

sure!

Reply:
following our discussion in the call, I understand that we rely on `bad_pmf` for DAU since we want to test the more specific error messages that you defined. I am wondering if we should even replace `bad_pmf_common` with `bad_pmf`: even for other methods, I assume that we would implement the more informative errors. what do you think? otherwise i am also fine to leave it as it is for now.

Reply:
> even for other methods, I assume that we would implement the more informative errors. what do you think?

It is possible that some UNU.RAN methods support infinite domains, and evaluating the PMF early on in such cases is not possible; we will have to rely on UNU.RAN and hope that it does everything correctly. In that case, we could use this list. But I will just remove it until needed. If we ever write methods that do require the list, we can reintroduce it.

Reply:
I have removed the list now.

Reply:
ok, I understand the difference better now. let's see if we need to add it later again